ccl/kvccl/kvfollowerreadsccl: TestFollowerReadsWithStaleDescriptor failed

ccl/kvccl/kvfollowerreadsccl.TestFollowerReadsWithStaleDescriptor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12832589?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12832589?buildTab=artifacts#/) on master @ [caff15394fcbe37208b46b2973714c27cc3a1417](https://github.com/cockroachdb/cockroach/commits/caff15394fcbe37208b46b2973714c27cc3a1417):


```
        	            	    28.169ms      0.143ms                === operation:dist sender send _verbose:1 node:3 f:689776af distsql.stmt:SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp() distsql.gateway:4 distsql.txn:049b8c56-a399-4112-b9fb-427f21347fa3 txn:049b8c56
        	            	    28.169ms      0.000ms                [/cockroach.roachpb.Internal/Batch: {count: 3, duration 8ms}]
        	            	    28.356ms      0.187ms                event:kv/kvclient/kvcoord/range_iter.go:188 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] querying next range at /Table/106/1/3/0
        	            	    28.739ms      0.382ms                event:kv/kvclient/kvcoord/range_iter.go:225 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] key: /Table/106/1/3/0, desc: r67:/{Table/65-Max} [(n1,s1):1, (n3,s3):3NON_VOTER, next=4, gen=6]
        	            	    28.862ms      0.124ms                event:kv/kvclient/kvcoord/dist_sender.go:2306 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] routing to nearest replica; leaseholder not required
        	            	    29.039ms      0.177ms                event:kv/kvclient/kvcoord/dist_sender.go:2364 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] r67: sending batch 1 Get to (n3,s3):3NON_VOTER
        	            	    29.105ms      0.066ms                event:rpc/nodedialer/nodedialer.go:138 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] sending request to local client
        	            	    29.165ms      0.060ms                event:kv/kvclient/kvcoord/transport.go:210 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] sending batch request
        	            	    29.219ms      0.054ms                    === operation:/cockroach.roachpb.Internal/Batch _verbose:1 node:3 span.kind:server request:Get [/Table/106/1/3/0,/Min), [txn: 049b8c56], [max_span_request_keys: 0], [target_bytes: 10485760]
        	            	    29.658ms      0.440ms                    event:server/node.go:1316 [n3] node received request: 1 Get
        	            	    30.067ms      0.408ms                    event:kv/kvserver/store_send.go:154 [n3,s3] executing Get [/Table/106/1/3/0,/Min), [txn: 049b8c56], [max_span_request_keys: 0], [target_bytes: 10485760]
        	            	    30.503ms      0.436ms                    event:kv/kvserver/replica_send.go:183 [n3,s3,r67/3:/{Table/65-Max}] read-only path
        	            	    30.667ms      0.164ms                    event:kv/kvserver/concurrency/concurrency_manager.go:209 [n3,s3,r67/3:/{Table/65-Max}] sequencing request
        	            	    30.747ms      0.080ms                    event:kv/kvserver/concurrency/concurrency_manager.go:290 [n3,s3,r67/3:/{Table/65-Max}] acquiring latches
        	            	    30.887ms      0.140ms                    event:kv/kvserver/concurrency/concurrency_manager.go:334 [n3,s3,r67/3:/{Table/65-Max}] scanning lock table for conflicting locks
        	            	    31.519ms      0.632ms                    event:kv/kvserver/replica_follower_read.go:118 [n3,s3,r67/3:/{Table/65-Max}] can't serve follower read; closed timestamp too low by: 19.004456ms; maxClosed: 1700896979.706059544,0 ts: 1700896979.725064000,0 uncertaintyLimit: 1700896979.725064000,0
        	            	    31.940ms      0.420ms                    event:kv/kvserver/replica_send.go:202 [n3,s3,r67/3:/{Table/65-Max}] replica.Send got error: [NotLeaseHolderError] lease held by different store; r67: replica (n3,s3):3NON_VOTER not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1700896978.781646021,0
        	            	    32.239ms      0.299ms                    event:server/node.go:1348 [n3] error from stores.Send: [NotLeaseHolderError] lease held by different store; r67: replica (n3,s3):3NON_VOTER not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1700896978.781646021,0: "sql txn" meta={id=049b8c56 key=/Min iso=Serializable pri=0.00974014 epo=0 ts=1700896979.725064000,0 min=1700896979.725064000,0 seq=0} lock=false stat=PENDING rts=1700896979.725064000,0 wto=false gul=1700896979.725064000,0
        	            	    32.878ms      0.639ms                    event:server/node.go:1382 [n3] node sending response
        	            	    33.198ms      4.033ms                event:kv/kvclient/kvcoord/transport.go:212 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] received batch response
        	            	    33.661ms      0.463ms                event:kv/kvclient/kvcoord/dist_sender.go:2520 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] [NotLeaseHolderError] lease held by different store; r67: replica (n3,s3):3NON_VOTER not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1700896978.781646021,0: "sql txn" meta={id=049b8c56 key=/Min iso=Serializable pri=0.00974014 epo=0 ts=1700896979.725064000,0 min=1700896979.725064000,0 seq=0} lock=false stat=PENDING rts=1700896979.725064000,0 wto=false gul=1700896979.725064000,0
        	            	    33.937ms      0.276ms                event:kv/kvclient/kvcoord/dist_sender.go:2666 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] application error: [NotLeaseHolderError] lease held by different store; r67: replica (n3,s3):3NON_VOTER not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1700896978.781646021,0: "sql txn" meta={id=049b8c56 key=/Min iso=Serializable pri=0.00974014 epo=0 ts=1700896979.725064000,0 min=1700896979.725064000,0 seq=0} lock=false stat=PENDING rts=1700896979.725064000,0 wto=false gul=1700896979.725064000,0
        	            	    34.010ms      0.073ms                event:kv/kvclient/kvcoord/dist_sender.go:2367 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] trying next peer (n1,s1):1
        	            	    34.080ms      0.070ms                event:rpc/nodedialer/nodedialer.go:147 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] sending request to 127.0.0.1:33021
        	            	    34.205ms      0.125ms                event:kv/kvclient/kvcoord/transport.go:210 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] sending batch request
        	            	    34.340ms      0.135ms                    === operation:/cockroach.roachpb.Internal/Batch _verbose:1 node:3 f:689776af distsql.stmt:SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp() distsql.gateway:4 distsql.txn:049b8c56-a399-4112-b9fb-427f21347fa3 txn:049b8c56 span.kind:client
        	            	    34.750ms      0.410ms                    === operation:/cockroach.roachpb.Internal/Batch _verbose:1 node:1 span.kind:server request:Get [/Table/106/1/3/0,/Min), [txn: 049b8c56], [max_span_request_keys: 0], [target_bytes: 10485760]
        	            	    34.840ms      0.090ms                    event:server/node.go:1316 [n1] node received request: 1 Get
        	            	    34.982ms      0.142ms                    event:kv/kvserver/store_send.go:154 [n1,s1] executing Get [/Table/106/1/3/0,/Min), [txn: 049b8c56], [max_span_request_keys: 0], [target_bytes: 10485760]
        	            	    35.074ms      0.092ms                    event:kv/kvserver/replica_send.go:183 [n1,s1,r67/1:/{Table/65-Max}] read-only path
        	            	    35.153ms      0.079ms                    event:kv/kvserver/concurrency/concurrency_manager.go:209 [n1,s1,r67/1:/{Table/65-Max}] sequencing request
        	            	    35.191ms      0.038ms                    event:kv/kvserver/concurrency/concurrency_manager.go:290 [n1,s1,r67/1:/{Table/65-Max}] acquiring latches
        	            	    35.250ms      0.058ms                    event:kv/kvserver/concurrency/concurrency_manager.go:334 [n1,s1,r67/1:/{Table/65-Max}] scanning lock table for conflicting locks
        	            	    35.466ms      0.217ms                    event:kv/kvserver/replica_read.go:302 [n1,s1,r67/1:/{Table/65-Max}] can drop latches early for batch (Get [/Table/106/1/3/0,/Min), [txn: 049b8c56], [max_span_request_keys: 0], [target_bytes: 10485760]); scanning lock table first to detect conflicts
        	            	    35.564ms      0.098ms                    event:kv/kvserver/replica_read.go:114 [n1,s1,r67/1:/{Table/65-Max}] lock table scan complete without conflicts; dropping latches early
        	            	    35.671ms      0.106ms                    event:kv/kvserver/replica_read.go:414 [n1,s1,r67/1:/{Table/65-Max}] executing read-only batch
        	            	    35.846ms      0.175ms                    event:kv/kvserver/replica_evaluate.go:508 [n1,s1,r67/1:/{Table/65-Max}] evaluated Get command header:<key:"\362\211\213\210" > , txn="sql txn" meta={id=049b8c56 key=/Min iso=Serializable pri=0.00974014 epo=0 ts=1700896979.725064000,0 min=1700896979.725064000,0 seq=0} lock=false stat=PENDING rts=1700896979.725064000,0 wto=false gul=1700896979.725064000,0 : resp=header:<> , err=<nil>
        	            	    35.925ms      0.079ms                    event:scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 1 gets, 0 scans, 0 reverse scans
        	            	    36.008ms      0.083ms                    event:kv/kvserver/replica_read.go:227 [n1,s1,r67/1:/{Table/65-Max}] read completed
        	            	    36.198ms      0.190ms                    event:server/node.go:1382 [n1] node sending response
        	            	    35.869ms     -0.329ms                    structured:{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","numInterfaceSteps":"0","numInternalSteps":"0","blockBytes":"0","blockBytesInCache":"0","keyBytes":"0","valueBytes":"0","pointCount":"0","pointsCoveredByRangeTombstones":"0","rangeKeyCount":"0","rangeKeyContainedPoints":"0","rangeKeySkippedPoints":"0","separatedPointCount":"0","separatedPointValueBytes":"0","separatedPointValueBytesFetched":"0","numGets":"1","numScans":"0","numReverseScans":"0"}
        	            	    36.866ms      2.661ms                event:kv/kvclient/kvcoord/transport.go:212 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] received batch response
        	            	    37.278ms      9.252ms            event:kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:644 [n3,f689776af,distsql.stmt=SELECT * FROM [SELECT * FROM test WHERE k = _ UNION ALL SELECT * FROM test WHERE k = _] AS OF SYSTEM TIME follower_read_timestamp(),distsql.gateway=4,distsql.txn=049b8c56-a399-4112-b9fb-427f21347fa3,txn=049b8c56] recording span to refresh: /Table/106/1/3/0
        	            	    41.087ms     28.807ms    event:sql/conn_executor_exec.go:1104 [n4,client=127.0.0.1:36690,hostssl,user=root] execution ends
        	            	    41.137ms      0.051ms    event:sql/conn_executor_exec.go:1104 [n4,client=127.0.0.1:36690,hostssl,user=root] rows affected: 0
        	            	    41.699ms      0.562ms        === operation:commit sql txn _verbose:1 node:4 client:127.0.0.1:36690 hostssl: user:root
        	            	    42.010ms      0.872ms    event:sql/conn_executor_exec.go:3012 [n4,client=127.0.0.1:36690,hostssl,user=root] AutoCommit. err: <nil>
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/becd5e00c91ac217b69adea284b6c9ea/logTestFollowerReadsWithStaleDescriptor789423768
--- FAIL: TestFollowerReadsWithStaleDescriptor (11.83s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFollowerReadsWithStaleDescriptor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>



Jira issue: CRDB-33864

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
@@ -801,12 +801,11 @@ func TestFollowerReadsWithStaleDescriptor(t *testing.T) {
 	// the following settings.
 	n1.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.target_duration = '0.1s'`)
 	n1.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '0.1s'`)
-	n1.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.1s'`)
 
 	// Sleep so that we can perform follower reads. The read timestamp needs to be
 	// above the timestamp when the table was created.
 	log.Infof(ctx, "test sleeping for the follower read timestamps to pass the table creation timestamp...")
-	time.Sleep(300 * time.Millisecond)
+	n1.Exec(t, `SELECT pg_sleep((now() - follower_read_timestamp())::FLOAT)`)
 	log.Infof(ctx, "test sleeping... done")
 
 	// Run a query on n4 to populate its cache.
@@ -1008,7 +1007,6 @@ func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 			systemSQL := sqlutils.MakeSQLRunner(tc.Conns[0])
 			systemSQL.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.target_duration = '0.1s'`)
 			systemSQL.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '0.1s'`)
-			systemSQL.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.1s'`)
 			// We're making assertions on traces collected by the tenant using
 			// log lines in KV so we must ensure they're not redacted.
 			systemSQL.Exec(t, `SET CLUSTER SETTING trace.redact_at_virtual_cluster_boundary.enabled = 'false'`)
@@ -1087,7 +1085,7 @@ func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 			// Wait until all tenant servers are aware of the setting override.
 			testutils.SucceedsSoon(t, func() error {
 				settingNames := []string{
-					"kv.closed_timestamp.target_duration", "kv.closed_timestamp.side_transport_interval", "kv.closed_timestamp.propagation_slack",
+					"kv.closed_timestamp.target_duration", "kv.closed_timestamp.side_transport_interval",
 				}
 				for _, settingName := range settingNames {
 					for i := 0; i < numNodes; i++ {
@@ -1130,7 +1128,7 @@ func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 			// Sleep so that we can perform follower reads. The read timestamp
 			// needs to be above the timestamp when the table was created.
 			log.Infof(ctx, "test sleeping for the follower read timestamps to pass the table creation timestamp...")
-			time.Sleep(500 * time.Millisecond)
+			tenantSQL.Exec(t, `SELECT pg_sleep((now() - follower_read_timestamp())::FLOAT)`)
 			log.Infof(ctx, "test sleeping... done")
 
 			// Check that the cache was indeed populated.

